src/jit/flowgraph.cpp

```diff
@@ -2759,87 +2679,83 @@ void Compiler::fgDispDomTree(BasicBlockList** domTree)
 #endif // DEBUG
 
 //------------------------------------------------------------------------
-// fgTraverseDomTree: Assign pre/post-order numbers to the dominator tree.
+// fgNumberDomTree: Assign pre/post-order numbers to the dominator tree.
 //
 // Arguments:
-//    bbNum   - The basic block number of the starting block
-//    domTree - The dominator tree (as child block lists)
-//    preNum  - Pointer to the pre-number counter
-//    postNum - Pointer to the post-number counter
+//    domTree - The dominator tree node array
 //
 // Notes:
-//    Runs a non-recursive DFS traversal of the dominator tree using an
-//    evaluation stack to assign pre-order and post-order numbers.
-//    These numberings are used to provide constant time lookup for
-//    ancestor/descendent tests between pairs of nodes in the tree.
-
-void Compiler::fgTraverseDomTree(unsigned bbNum, BasicBlockList** domTree, unsigned* preNum, unsigned* postNum)
+//    Runs a non-recursive DFS traversal of the dominator tree to assign
+//    pre-order and post-order numbers. These numbers are used to provide
+//    constant time lookup ancestor/descendent tests between pairs of nodes
+//    in the tree.
+//
+void Compiler::fgNumberDomTree(DomTreeNode* domTree)
 {
-    noway_assert(bbNum <= fgBBNumMax);
-
-    // If the block preorder number is not zero it means we already visited
-    // that node, so we skip it.
-    if (fgDomTreePreOrder[bbNum] == 0)
+    class NumberDomTreeVisitor : public DomTreeVisitor<NumberDomTreeVisitor>
     {
-        // If this is the first time we visit this node, both preorder and postnumber
-        // values must be zero.
-        noway_assert(fgDomTreePostOrder[bbNum] == 0);
+        unsigned m_preNum;
+        unsigned m_postNum;
 
-        // Allocate a local stack to hold the Dfs traversal actions necessary
-        // to compute pre/post-ordering of the dominator tree.
-        ArrayStack<DfsNumEntry> stack(getAllocator(CMK_ArrayStack));
-
-        // Push the first entry number on the stack to seed the traversal.
-        stack.Push(DfsNumEntry(DSS_Pre, bbNum));
+    public:
+        NumberDomTreeVisitor(Compiler* compiler, DomTreeNode* domTree) : DomTreeVisitor(compiler, domTree)
+        {
+        }
 
-        // The search is terminated once all the actions have been processed.
-        while (!stack.Empty())
+        void Begin()
         {
-            DfsNumEntry current    = stack.Pop();
-            unsigned    currentNum = current.dfsNum;
+            unsigned bbArraySize           = m_compiler->fgBBNumMax + 1;
+            m_compiler->fgDomTreePreOrder  = new (m_compiler, CMK_DominatorMemory) unsigned[bbArraySize];
```

Comment:
`new (m_compiler, CMK_DominatorMemory) unsigned[bbArraySize]()` for value-initialization or `unsigned[bbArraySize]{}` for list-initialization and delete `// Initialize all the data structures.` block?

Reply:
Right, I think I avoided this kind of initialization because I vaguely remember an old version of MSVC used by internal JIT builds having a bug in this area. But hopefully that's history now.

Also, I should add a constructor to `DomTreeNode` to avoid manual initialization as well.